lib/auth/periodic_test.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package auth implements certificate signing authority and access control server
+// Authority server is composed of several parts:
+//
+// * Authority server itself that implements signing and acl logic
+// * HTTP server wrapper for authority server
+// * HTTP client wrapper
+package auth
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+func TestInstanceMetricsPeriodic(t *testing.T) {
+	tts := []struct {
+		desc           string
+		upgraders      []string
+		expectCounts   map[string]int
+		expectEnrolled int
+	}{
+		{
+			desc: "mixed",
+			upgraders: []string{
+				"kube",
+				"unit",
+				"",
+				"unit",
+				"",
+			},
+			expectCounts: map[string]int{
+				"kube": 1,
+				"unit": 2,
+			},
+			expectEnrolled: 3,
+		},
+		{
+			desc: "all-unenrolled",
+			upgraders: []string{
+				"",
+				"",
+			},
+		},
+		{
+			desc: "all-enrolled",
+			upgraders: []string{
+				"kube",
+				"kube",
+				"unit",
+				"unit",
+			},
+			expectCounts: map[string]int{
+				"kube": 2,
+				"unit": 2,
+			},
+			expectEnrolled: 4,
+		},
+		{
+			desc: "nothing",
+		},
+	}
+
+	for _, tt := range tts {
+		periodic := newInstanceMetricsPeriodic()
```

Comment:
Here and below would it be worth using subtests with `t.Run` so that we can run or skip individual test cases?